doc/hacking/style_guide.md

```diff
@@ -0,0 +1,187 @@
+Panda3D C++ Style Guide
+=======================
+
+Rather than try to justify every decision made, this document attempts to be as
+at-a-glance as possible.  Examples provided where necessary to illustrate best
+practices.
+
+Note that this style guide is, for the most part, not very strict, and there
+are exceptions to every rule.  However, this guide does serve to illustrate the
+preferred style for the entire codebase.
+
+Naming conventions
+------------------
+
+- **Classes**: `CamelCaseFirstUpper`
+- **Filenames**: `lowerCamelCase.h`, `lowerCamelCase.cxx`
+  - Note, every header file must have a unique filename.  They are all
+    installed into the same path.
+- **Functions**: `lower_snake_case_without_prefix`
+- **Local variables**: `lower_snake_case_without_prefix`
+- **Member variables**: `_lower_snake_case_with_initial_score`
```

Comment:
I also would like to see enum classes. They are type safe, and less error-prone to use. I think we should refer to enum classes as the desired style (which does not mean we have to convert all old code - just to make sure new code uses this convention)

Reply:
More about enums: 

 1. Spacing for enum values? (Currently we seem to use 1))
```cpp
enum class MyEnum {
  my_value = 1, // 1)
  other_value=2 // or 2) ?
}
```

2. Naming for enum values. Assuming we use enum classes, how would the members be named? `unsigned_byte`, `UnsignedByte` or `unsignedByte` ? I think the current style would be `unsigned_byte` then.

Reply:
I agree on using enum classes in the future.

I think enum values should be spaced because they look like assignments, but I can go either way on that.

As for enumerant naming style, I would suggest `unsigned_byte` or `UNSIGNED_BYTE`.  Normally I would strongly prefer the former, but it seems that [Python settled on](https://mail.python.org/pipermail/python-ideas/2016-September/042340.html) using `UNSIGNED_BYTE` for enumerants (and all other constants), so I see three options:
1. Defy the Python convention and use the lowercase style.
2. Use the UPPER_CASE style both in Python and C++ code.
3. Use lower_case in C++, but let interrogate map this to UPPER_CASE.

Reply:
@rdb I think python is a special case, since you usually declare and access them like class members, so its nice to distinguish. Since in C++, one would have to write `Texture::Format::unsigned_byte` I think it would be better to stick to underscores, since `Texture::Format::UNSIGNED_BYTE` looks weird to me (Just my opinion)

For the python interface we could probably make them upper case then.